### PR TITLE
fix: add redirect_after_logout_uri for ODIC that do not have an end_session_endpoint

### LINKED ISSUE #6791: bug: openid-connect plugin doesn't redirect after logout
### Current Behavior

I would like to integrate APISIX gateway with [auth0](https://auth0.com) using [openid-connect](https://apisix.apache.org/docs/apisix/plugins/openid-connect/) plugin.
After a request is sent to `/logout`, the session is invalidated on the server side and `session=` field in the cookie set to empty on the client side. In addition to that a request needs be sent to auth0's logout endpoint ( see [here](https://auth0.com/docs/authenticate/login/logout/log-users-out-of-auth0) ). So, `post_logout_redirect_uri` field is configured in the plugin to make logout request to the auth0 endpoint, but it doesn't redirect. 

### Expected Behavior

When `post_logout_redirect_uri` is configured in openid-connect plugin, it should redirect user to the given URI after a request sent to the `/logout` endpoint on the gateway. 

### Error Logs

No errors printed to `/usr/local/apisix/logs/error.log`

### Steps to Reproduce

1. Run APISIX via [the Docker deployment](https://github.com/apache/apisix-docker)
```
cd example
docker-compose -p docker-apisix up -d
```

2. Create a Route in the dashboard
3. Enable/Configure openid-connect plugin (Scope: global). 
```
{
  "access_token_in_authorization_header": true,
  "bearer_only": false,
  "client_id": "{client_id}",
  "client_secret": "{client_secret}",
  "disable": false,
  "discovery": "https://{domain}.us.auth0.com/.well-known/openid-configuration",
  "introspection_endpoint_auth_method": "client_secret_post",  
  "realm": "master",
  "redirect_uri": "http://127.0.0.1:9080/",
  "scope": "openid profile",
  "post_logout_redirect_uri": "https://{domain}.us.auth0.com/v2/logout?client_id={client_id}&returnTo=http://127.0.0.1:9080/app",
}
```
4. Authenticate user by going to any defined route on the gateway ( e.g. `/protected` ). It will redirect user to the auth0 login page .
5. Go to `/logout`, no redirection happens.

Screenshot from 2022-04-04 12-41-29

### Environment

- APISIX version (run `apisix version`): **2.13.0**
- Operating system (run `uname -a`): the apisix docker image
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`):  nginx version: openresty/1.19.9.1
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`): the apisix docker image
- APISIX Dashboard version, if relevant: **2.10.1**
- Plugin runner version, for issues related to plugin runners: -
- LuaRocks version, for installation issues (run `luarocks --version`): -

### PR DESCRIPTION
**Pull request by luoluoyuyu**
### Description

Fixes #6791
#### Problem Analysis
 When the openid service does not support `end_session_endpoint`, the ODIC logout  fails to redirect to post_logout_redirect_uri. this issue is caused by the [OpenID]( https://openid.net/specs/openid-connect-rpinitiated-1_0.html#OPMetadata) standard. `end_session_endpoint` does not necessarily need to be implemented.

#### Solution
Use the `redirect_after_logout_uri` defined in [lua-resty-openidc](https://github.com/zmartzone/lua-resty-openidc/blob/master/lib/resty/openidc.lua#L1312-L1326) for redirection when the OpenID service does not support `end_session_endpoint`.

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by Sn0rt** [comment]
1. new test for cover
2. user doc

**Comment by luoluoyuyu** [comment]
The solution to the problem is to use `redirect_after_logout_uri` for redirection after logout if there is no `end_session_endpoint` in the discovery metadata, but note that since `redirect_after_logout_uri` is not defined in the schema, it can only be configured in the rewrite phase.  If you set the value of `redirect_after_logout_uri` in the check_schema phase, you will get an error when restarting apisix.

**Comment by luoluoyuyu** [comment]
It's not necessary to add documentation, this PR is just to support services like Google that don't provide the [end_session_endpoint](https://accounts.google.com/.well-known/openid-configuration) configuration, and OIDC can also correctly using the `post_logout_redirect_uri` configuration.

**Comment by luoluoyuyu** [comment]
> new test for cover

Auth0 now supports `end_session_endpoint`, now only found google openid service does not support `end_session_endpoint`, if you want to add test cases for google need to google account password, I think add test cases is not appropriate.
Here is the OpenID service that supports `end_session_endpoint` that I have organized:
Okta: https://developer.okta.com/docs/reference/api/oidc/#response-properties-10
Azure: https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols-oidc#single-sign-out
Auth0: https://auth0.com/docs/authenticate/login/logout/log-users-out-of-auth0#endpoint-discovery
Curity: https://curity.io/resources/learn/openid-connect-logout/#rp-initiated-logout
Connect2id: https://connect2id.com/products/nimbus-oauth-openid-connect-sdk#specs
Gluu: https://gluu.org/docs/gluu-server/4.3/api-guide/openid-connect-api/#openid-connect-end-session-api
IdentityServer: https://docs.duendesoftware.com/identityserver/v6/ui/logout/session_cleanup/#prompting-the-user-to-logout

File: apisix/plugins/openid-connect.lua [hunk 1]
```diff
@@ -440,15 +440,30 @@ function _M.rewrite(plugin_conf, ctx)
...
     if not conf.ssl_verify then
         -- openidc use "no" to disable ssl verification
         conf.ssl_verify = "no"
     end
 
+    if path == (conf.logout_path or "/logout") then
+        local discovery, discovery_err =openidc.get_discovery_doc(conf)
+        if discovery_err then
+            core.log.error("OIDC access discovery url failed : ", discovery_err)
+            return 500
```
Comment (monkeyDluffy6017) [inline apisix/plugins/openid-connect.lua:458]: Why 500 here? 500 means `internal error`, we should not return 500 on any occasion

Comment (luoluoyuyu) [reply apisix/plugins/openid-connect.lua:458]: The reason for returning 500 is the same as the reason for returning 500 in this step. https://github.com/apache/apisix/blob/master/apisix/plugins/openid-connect.lua#L507-L518。
The `openidc.authenticate` and `get_discovery_doc` functions both get discovered metadata via openidc_ensure_discovered_data.https://github.com/zmartzone/lua-resty-openidc/blob/master/lib/resty/openidc.lua#L580

File: apisix/plugins/openid-connect.lua [hunk 2]
```diff
@@ -440,15 +440,30 @@ function _M.rewrite(plugin_conf, ctx)
...
+        conf.redirect_uri = path
     end
 
     if not conf.ssl_verify then
         -- openidc use "no" to disable ssl verification
         conf.ssl_verify = "no"
     end
 
+    if path == (conf.logout_path or "/logout") then
+        local discovery, discovery_err =openidc.get_discovery_doc(conf)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/openid-connect.lua:455]:
```suggestion
        local discovery, discovery_err = openidc.get_discovery_doc(conf)
```

File: apisix/plugins/openid-connect.lua [hunk 3]
```diff
@@ -504,7 +519,7 @@ function _M.rewrite(plugin_conf, ctx)
         -- provider's authorization endpoint to initiate the Relying Party flow.
         -- This code path also handles when the ID provider then redirects to
         -- the configured redirect URI after successful authentication.
-        response, err, _, session  = openidc.authenticate(conf, nil, unauth_action, conf.session)
+        response, err, _, session  = openidc.authenticate(conf, path, unauth_action, conf.session)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/openid-connect.lua:522]: Why do you change this?

Comment (luoluoyuyu) [reply apisix/plugins/openid-connect.lua:522]: It's just for the code to be better understood, it doesn't change the code logic, the default used is also ngx.var.request_uri.   https://github.com/zmartzone/lua-resty-openidc/blob/master/lib/resty/openidc.lua#L1461

Comment (moonming) [reply apisix/plugins/openid-connect.lua]: Do not make irrelevant changes

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by moonming — CHANGES_REQUESTED** [review]
can not merge without test cases

**Review by monkeyDluffy6017 — APPROVED** [review]

File: t/plugin/openid-connect.t [hunk 4]
```diff
@@ -1303,3 +1303,125 @@ passed
...
+            end
+            ngx.say(body)
+        }
+    }
+--- response_body
+passed
+
+
+
+=== TEST 35: Check whether outh0 can redirect normally using post_logout_redirect_uri configuration
```
Comment (moonming) [inline t/plugin/openid-connect.t:1347]: auth0

**Review by moonming — CHANGES_REQUESTED** [review]

**Review by moonming — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]
great job

**Review by kayx23 — APPROVED** [review]

**Comment by kayx23** [comment]
@luoluoyuyu a question for the below:

image

What if someone configures both `post_logout_redirect_uri` and `redirect_after_logout_uri`?

https://github.com/zmartzone/lua-resty-openidc/blob/master/tests/spec/logout_spec.lua#L264-L299

**Comment by luoluoyuyu** [comment]
> @luoluoyuyu a question for the below:
> 
> image
> 
> What if someone configures both `post_logout_redirect_uri` and `redirect_after_logout_uri`?
> 
> https://github.com/zmartzone/lua-resty-openidc/blob/master/tests/spec/logout_spec.lua#L264-L299

Hi @kayx23  Deng 'redirect_after_logout_uri' is not defined in the schema. If 'redirect_after_logout_uri' is configured, an error will occur. [apisix/apisix/plugins/openid-connect.lua at master · apache/apisix (github.com)](https://github.com/apache/apisix/blob/master/apisix/plugins/openid-connect.lua#L32C17-L33)

**Comment by kayx23** [comment]
>  If 'redirect_after_logout_uri' is configured, an error will occur.

Not really true as previously shown.

image

However I will remove `redirect_after_logout_uri` from the new doc given this change.

**Pull request merged**